setuptools/config/pyprojecttoml.py

```diff
@@ -365,6 +374,37 @@ def _obtain_classifiers(self, dist: "Distribution"):
                 return value.splitlines()
         return None
 
+    def _obtain_dependencies(self, dist: "Distribution"):
+        if "dependencies" in self.dynamic:
+            value = self._obtain(dist, "dependencies", {})
+            if value:
+                return _parse_requirements_list(value)
+        return None
+
+    def _obtain_optional_dependencies(self, dist: "Distribution"):
+        if "optional-dependencies" in self.dynamic:
+            optional_dependencies_map = self.dynamic_cfg["optional-dependencies"]
+            assert isinstance(optional_dependencies_map, dict)
+            expanded_optional_dependencies_map = {
+                group: _parse_requirements_list(self._expand_directive(
+                    f"tool.setuptools.dynamic.optional-dependencies.{group}",
+                    directive,
+                    {},
+                ))
+                for group, directive in optional_dependencies_map.items()
+            }
+            self._ensure_previously_set(dist, "optional-dependencies")
+            return expanded_optional_dependencies_map
```

Comment:
Sorry, I was just seeing the CI errors, and it seems to be a bit more complicated than I first thought. How about?

Sorry, according to the error in the CI, it seems a bit more complicated than that...
How about:

```python
    def _obtain_optional_dependencies(self, dist: "Distribution"):
        if "optional-dependencies" not in self.dynamic:
            return None
        if "optional-dependencies" in self.dynamic_cfg:
            optional_dependencies_map = self.dynamic_cfg["optional-dependencies"]
            ...
       self._ensure_previously_set(dist, "optional-dependencies")
       return None
```

Reply:
Here `self._ensure_previously_set` + `return None` are the fallback code path that require the user to specify `extras_require` via `setup.py` if `dynamic` contains `optional-dependencies`.